src/imitation/algorithms/mce_irl.py

```diff
@@ -321,49 +321,92 @@ def __init__(
             rng=self.rng,
         )
 
-    def _set_demo_from_obs(self, obses: np.ndarray) -> None:
-        if self.discount != 1.0:
-            raise ValueError(
-                "Cannot compute discounted OM from timeless Transitions.",
-            )
+    def _set_demo_from_trajectories(self, trajs: Iterable[types.Trajectory]) -> None:
+        num_demos = 0
+        for traj in trajs:
+            cum_discount = 1.0
+            for obs in traj.obs:
+                self.demo_state_om[obs] += cum_discount
+                cum_discount *= self.discount
+            num_demos += 1
+        self.demo_state_om /= num_demos
+
+    def _update_demo_from_obs(
+        self,
+        obses: np.ndarray,
+        dones: Optional[np.ndarray],
+        next_obses: Optional[np.ndarray],
+    ) -> None:
         for obs in obses:
             if isinstance(obs, th.Tensor):
                 obs = obs.numpy()
-            self.demo_state_om[obs.astype(bool)] += 1.0
+            self.demo_state_om[obs] += 1.0
+
+        # We assume the transitions were flattened from some trajectories,
+        # then possibly shuffled. So add next observations for terminal states,
+        # as they will not appear anywhere else; but ignore next observations
+        # for all other states as they occur elsewhere in dataset.
+        if next_obses is not None:
+            for done, obs in zip(dones, next_obses):
+                if isinstance(done, th.Tensor):
+                    done = done.numpy()
```

Comment:
Nit/defensive programming:

```suggestion
                    done = done.item()  # must be scalar
```